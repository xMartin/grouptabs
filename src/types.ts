export interface ActionMap {
  createOrUpdate: Entity[];
  delete: Entity[];
}

export enum DocumentType {
  TRANSACTION = "transaction",
  INFO = "info",
}

export enum TransactionType {
  SHARED = "SHARED",
  DIRECT = "DIRECT",
}

export interface Account {
  participant: string;
  amount: number;
}

export interface Transaction {
  id: string;
  type: DocumentType.TRANSACTION;
  tabId: string;
  timestamp: string;
  transactionType: TransactionType;
  description: string;
  date: string;
  participants: Account[];
}

export interface Info {
  id: string;
  type: DocumentType.INFO;
  tabId: string;
  name: string;
}

export type Entity = Transaction | Info;

export interface Tab {
  id: string;
  info: Info;
  mostRecentTransaction?: Transaction;
}

export enum TransactionFormParticipantStatus {
  NONE = 0,
  JOINED = 1,
  PAID = 2,
}

export enum TransactionFormParticipantInputType {
  NEW,
  EXISTING,
}

export interface TransactionFormSharedState {
  id: string;
  inputType: TransactionFormParticipantInputType;
  participant?: string;
  status: TransactionFormParticipantStatus;
  amount?: number;
}

interface TransactionFormDirectState {
  from?: string;
  fromNew?: string;
  to?: string;
  toNew?: string;
  amount?: number;
  options: string[];
}

export interface TransactionFormState {
  transactionType: TransactionType;
  date: string;
  description?: string;
  shared: TransactionFormSharedState[];
  direct: TransactionFormDirectState;
}
