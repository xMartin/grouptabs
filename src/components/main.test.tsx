import renderer from "react-test-renderer";
import Main from "./main";
import { DocumentType, TransactionType } from "../types";

it("renders empty view with no tab selected", () => {
  const tree = renderer
    .create(
      <Main
        accounts={[]}
        transactions={[]}
        total={0}
        onChangeTabClick={vi.fn()}
        onNavigateToAddTransaction={vi.fn()}
        onDetailsClick={vi.fn()}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders tab with no transactions", () => {
  const tree = renderer
    .create(
      <Main
        tabId="6"
        tabInfo={{
          id: "info-6",
          type: DocumentType.INFO,
          tabId: "6",
          name: "My Tab",
        }}
        accounts={[]}
        transactions={[]}
        total={0}
        onChangeTabClick={vi.fn()}
        onNavigateToAddTransaction={vi.fn()}
        onDetailsClick={vi.fn()}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders summary and transaction list", () => {
  const tree = renderer
    .create(
      <Main
        tabId="2"
        tabInfo={{
          id: "info-2",
          type: DocumentType.INFO,
          tabId: "2",
          name: "My Tab",
        }}
        accounts={[
          {
            participant: "Jan",
            amount: -11.7,
          },
          {
            participant: "Martin",
            amount: 11.7,
          },
        ]}
        transactions={[
          {
            id: "T4",
            tabId: "2",
            type: DocumentType.TRANSACTION,
            transactionType: TransactionType.SHARED,
            timestamp: "2020-09-19T23:59:59.999Z",
            date: "2020-09-20",
            description: "GummibÃ¤rchen",
            participants: [
              {
                participant: "Martin",
                amount: 22.8,
              },
              {
                participant: "Jan",
                amount: 0,
              },
            ],
          },
        ]}
        total={22.8}
        onChangeTabClick={vi.fn()}
        onNavigateToAddTransaction={vi.fn()}
        onDetailsClick={vi.fn()}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders missing tab info error", () => {
  const tree = renderer
    .create(
      <Main
        tabId="1"
        accounts={[]}
        transactions={[]}
        total={0}
        onChangeTabClick={vi.fn()}
        onNavigateToAddTransaction={vi.fn()}
        onDetailsClick={vi.fn()}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
