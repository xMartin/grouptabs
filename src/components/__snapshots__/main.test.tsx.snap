// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty view with no tab selected 1`] = `
<div
  className="scene mainScene"
>
  <div
    className="header"
  >
    <button
      className="left"
      onClick={[MockFunction]}
    >
      <svg
        height="16"
        width="16"
      >
        <path
          d="m2 2c-0.554 0-1 0.446-1 1s0.446 1 1 1h12c0.554 0 1-0.446 1-1s-0.446-1-1-1h-12zm0 5c-0.554 0-1 0.446-1 1s0.446 1 1 1h12c0.554 0 1-0.446 1-1s-0.446-1-1-1h-12zm0 5c-0.554 0-1 0.446-1 1s0.446 1 1 1h12c0.554 0 1-0.446 1-1s-0.446-1-1-1h-12z"
        />
      </svg>
    </button>
    <h2 />
    <button
      className="create"
      onClick={[Function]}
    >
      +
    </button>
  </div>
  <div
    className="content"
    id="main-content"
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <div
      className="load-error"
    >
      <img
        alt=""
        src="logo.png"
      />
      <h2>
        Grouptabs
      </h2>
      <p>
        Error: Tab data missing. Are you offline? Try refreshing.
      </p>
      <button
        className="create"
        onClick={[Function]}
      >
        OK
      </button>
    </div>
  </div>
</div>
`;

exports[`renders missing tab info error 1`] = `
<div
  className="scene mainScene"
>
  <div
    className="header"
  >
    <button
      className="left"
      onClick={[MockFunction]}
    >
      <svg
        height="16"
        width="16"
      >
        <path
          d="m2 2c-0.554 0-1 0.446-1 1s0.446 1 1 1h12c0.554 0 1-0.446 1-1s-0.446-1-1-1h-12zm0 5c-0.554 0-1 0.446-1 1s0.446 1 1 1h12c0.554 0 1-0.446 1-1s-0.446-1-1-1h-12zm0 5c-0.554 0-1 0.446-1 1s0.446 1 1 1h12c0.554 0 1-0.446 1-1s-0.446-1-1-1h-12z"
        />
      </svg>
    </button>
    <h2 />
    <button
      className="create"
      onClick={[Function]}
    >
      +
    </button>
  </div>
  <div
    className="content"
    id="main-content"
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <div
      className="load-error"
    >
      <img
        alt=""
        src="logo.png"
      />
      <h2>
        Grouptabs
      </h2>
      <p>
        Error: Tab data missing. Are you offline? Try refreshing.
      </p>
      <button
        className="create"
        onClick={[Function]}
      >
        OK
      </button>
    </div>
  </div>
</div>
`;

exports[`renders summary and transaction list 1`] = `
<div
  className="scene mainScene"
>
  <div
    className="header"
  >
    <button
      className="left"
      onClick={[MockFunction]}
    >
      <svg
        height="16"
        width="16"
      >
        <path
          d="m2 2c-0.554 0-1 0.446-1 1s0.446 1 1 1h12c0.554 0 1-0.446 1-1s-0.446-1-1-1h-12zm0 5c-0.554 0-1 0.446-1 1s0.446 1 1 1h12c0.554 0 1-0.446 1-1s-0.446-1-1-1h-12zm0 5c-0.554 0-1 0.446-1 1s0.446 1 1 1h12c0.554 0 1-0.446 1-1s-0.446-1-1-1h-12z"
        />
      </svg>
    </button>
    <h2>
      My Tab
    </h2>
    <button
      className="create"
      onClick={[Function]}
    >
      +
    </button>
  </div>
  <div
    className="content"
    id="main-content"
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <div
      className="row"
    >
      <table
        className="summary"
      >
        <tbody>
          <tr
            style={
              Object {
                "backgroundColor": "rgba(226,91,29,1)",
                "color": "var(--dark-color)",
              }
            }
          >
            <th
              className="account"
            >
              Jan
            </th>
            <td
              className="amount"
            >
              -11.7
            </td>
          </tr>
          <tr
            style={
              Object {
                "backgroundColor": "rgba(92,226,14,1)",
                "color": "var(--dark-color)",
              }
            }
          >
            <th
              className="account"
            >
              Martin
            </th>
            <td
              className="amount"
            >
              11.7
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      className="row"
    >
      <h3
        className="transactions-heading"
      >
        Transactions
      </h3>
      <div
        id="transactions"
      >
        <div>
          <div>
            <div
              className="dategroup"
            >
              9/20/2020
            </div>
            <div
              className="transaction list-item-with-arrow"
              onClick={[Function]}
            >
              <table>
                <tbody>
                  <tr>
                    <td
                      className="title"
                    >
                      Gummib√§rchen
                      <div
                        className="payments"
                      >
                        <strong>
                          Martin: 22.8, 
                        </strong>
                        Jan
                      </div>
                    </td>
                    <td
                      className="total"
                    >
                      22.8
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
      <div
        className="total-sum"
      >
        Total Spending: 
        22.8
      </div>
    </div>
    <div
      className="share-info"
    >
      <p>
        Share this tab ID for collaboration with others:
        <br />
        <code>
          2
        </code>
      </p>
    </div>
  </div>
</div>
`;

exports[`renders tab with no transactions 1`] = `
<div
  className="scene mainScene"
>
  <div
    className="header"
  >
    <button
      className="left"
      onClick={[MockFunction]}
    >
      <svg
        height="16"
        width="16"
      >
        <path
          d="m2 2c-0.554 0-1 0.446-1 1s0.446 1 1 1h12c0.554 0 1-0.446 1-1s-0.446-1-1-1h-12zm0 5c-0.554 0-1 0.446-1 1s0.446 1 1 1h12c0.554 0 1-0.446 1-1s-0.446-1-1-1h-12zm0 5c-0.554 0-1 0.446-1 1s0.446 1 1 1h12c0.554 0 1-0.446 1-1s-0.446-1-1-1h-12z"
        />
      </svg>
    </button>
    <h2>
      My Tab
    </h2>
    <button
      className="create"
      onClick={[Function]}
    >
      +
    </button>
  </div>
  <div
    className="content"
    id="main-content"
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <div
      className="empty-info"
    >
      <p>
        A tab consists of transactions. When you add a transaction you also define the people that are part of it, the participants.
      </p>
      <p>
        Start by adding your first transaction:
      </p>
      <div
        className="row"
      >
        <button
          className="full-width create"
          onClick={[Function]}
        >
          Add transaction
        </button>
      </div>
    </div>
    <div
      className="share-info"
    >
      <p>
        Share this tab ID for collaboration with others:
        <br />
        <code>
          6
        </code>
      </p>
    </div>
  </div>
</div>
`;
